### PR TITLE
fix(animations): convert numeric string values to numbers during animation params interpolation

### PR DESCRIPTION
**Pull request by dario-piotrowicz**
As I noticed here: https://github.com/angular/angular/pull/42763#issuecomment-877648810

Currently animation params get converted to strings, causing an error, where that could be avoided

For example, the developer can currently set a height of 100px in the component's decorator in following way:
	`style({ height: 100 })`
(100 is then automatically converted to `'100px'`)
but if they instead do
	`style({ height: '{{myHeight}}' })`
  with
	`[@animation]="{ value: '', params: { myHeight: 100 } }"`
  in the template,
they will get an error and the animation will not work, as currently `100` substitutes `{{myHeight}}` in the provided string, which becomes `'100'` which is invalid since it lacks the unit (px can be implied only if the value is numeric)

What I am doing is simply keeping `100` as a number so that it can get handled as a number by the rest of the process

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  &nbsp; &nbsp; &nbsp; :arrow_left:  (I thought it was a bug but I don't think it is)
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

As mentioned above providing numeric values as animations params doesn't work and yields an error

As you can see in this [stackblitz](https://stackblitz.com/edit/angular-ivy-xjni8d?file=src%2Fapp%2Fapp.component.html):

Peek 2021-08-15 00-19

Issue Number: N/A

## What is the new behavior?
Strings referring to a single numeric param get converted to numbers so that they can later be correctly handled

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

@JoostK Just tagging you in case you would like to have a look at this change since we discussed about this in the other PR :slightly_smiling_face: 

&NewLine;
&NewLine;
I didn't add unit tests for this, I am not totally sure this grants for new unit tests, please let me know if I should

&NewLine;
&NewLine;
By the way... as I worked on this I sort of realized something/changed my mind, the "issue" I am fixing is not really a real issue, since the developer is actually already using strings and thus should be aware of the nuances that come with that... 
 
I mean, the way it works now it's just natural and may seem unintuitive if you look at your param which may indeed be numbers but not if you look at your usage (which indeed strings)

So basically what I did is that if someone specifies something like `"{{ fontSize }}"` I actually convert that to a number, which then gets handled as such... but at the end of the day the user did provide a string...
 
So I am not completely sure, I think my change may be a nice fallback if the user actually intended to use the value as a number as I mentioned above (so that they won't have to specify `'px'`, just like how it is in other places) but it isn't like the current implementation is wrong or anything...

So yeah I think this may be a nice-to-have, but it's in no way necessary or anything...

what do you think?

### CODE REVIEW

File: packages/animations/browser/src/util.ts [hunk 1]
```diff
@@ -227,19 +227,31 @@ export function extractStyleParams(value: string|number): string[] {
 
 export function interpolateParams(
     value: string|number, params: {[name: string]: any}, errors: any[]): string|number {
+  if (typeof value === 'number') {
+    return value;
+  }
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:232]: if the value is a number then there is no need to perform any actions/checks so I thought it just made more sense to return it right away

File: packages/animations/browser/src/util.ts [hunk 2]
```diff
@@ -227,19 +227,31 @@ export function extractStyleParams(value: string|number): string[] {
 
 export function interpolateParams(
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:228]: I would have really liked to put a nice jsdoc comment here, but since no function in the file has one I thought I shouldn't as it would look out of place, but if you think I can please let me know and I get to it :slightly_smiling_face:

File: packages/animations/browser/src/util.ts [hunk 3]
```diff
@@ -227,19 +227,31 @@ export function extractStyleParams(value: string|number): string[] {
...
+        localVal = localVal.toString();
+      }
     }
-    return localVal.toString();
-  });
+  }
 
-  // we do this to assert that numeric values stay as they are
-  return str == original ? value : str;
+  return varName ? localVal : value;
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:254]: if there was a match for the regex then I use the newly created `localVal` otherwise I fall back to the original `value`

File: packages/animations/browser/src/util.ts [hunk 4]
```diff
@@ -227,19 +227,31 @@ export function extractStyleParams(value: string|number): string[] {
...
+  const match = PARAM_REGEX.exec(original);
+  const varName = match?.[1];
+  if (varName) {
     if (!params.hasOwnProperty(varName)) {
+      // this means that the value was never overridden by the data passed in by the user
       errors.push(`Please provide a value for the animation param ${varName}`);
       localVal = '';
+    } else {
+      localVal = params[varName];
+      // only numeric values are allowed not to be strings as they will be converted later on
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:247]: I added this comment because before everything was being converted into a string, but now I want to convert anything **except** numbers (and strings as they are already of the right type)

so for example if the param's value is `true` I convert it to `"true"`, etc...
I checked and this works the same as the current behaviour (except for numbers of course)

**Comment by dario-piotrowicz** [comment]
I converted this to draft as I noticed that the tests are failing and I am not handling correctly a situation such as this:
Screenshot at 2021-08-15 11-33-58

:cry: sorry, I'll need to rework a bit my solution :sweat_smile:

**Comment by dario-piotrowicz** [comment]
Should be all good now :slightly_smiling_face: 

(I've got a failing `test_aio_local` but I don't think that that is related to my changes in any way :thinking: )

**Comment by dario-piotrowicz** [comment]
By the way... I sort of realized something/changed my mind, the "issue" I am fixing is not really a real issue, since the developer is actually already using strings and thus should be aware of the nuances that come with that...

I mean, the way it works now it's just natural and may seem unintuitive if you look at your param which may indeed be numbers but not if you look at your usage (which indeed strings)

So basically what I did is that if someone specifies something like `"{{ fontSize }}"` I actually convert that to a number, which then gets handled as such... but at the end of the day the user did provide a string...

So I am not completely sure, I think my change may be a nice fallback if the user actually intended to use the value as a number as I mentioned above (so that they won't have to specify `'px'`, just like how it is in other places) but it isn't like the current implementation is wrong or anything...

So yeah I think this may be a nice-to-have, but it's in no way necessary or anything...

what do you think?

File: packages/animations/browser/src/util.ts [hunk 5]
```diff
@@ -211,8 +211,12 @@ export function validateStyleParams(
   }
 }
 
+const SINGLE_PARAM_REGEX = new RegExp(`^${SUBSTITUTION_EXPR_START}\\s*((?:(?!${
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:214]: only matches precise strings such as `{{ someVar }}` (with any number of spaces inside the brackets)

anything else won't get matched (and thus converted), for example `{{ someVar}} ` will not (and remain a string regardless on whether the param is a number or not)

File: packages/animations/browser/src/util.ts [hunk 6]
```diff
@@ -227,8 +231,18 @@ export function extractStyleParams(value: string|number): string[] {
...
+    return value;
+  }
+
   const original = value.toString();
-  const str = original.replace(PARAM_REGEX, (_, varName) => {
+
+  const matches = original.match(SINGLE_PARAM_REGEX);
+  if (matches) {
+    return interpolateSingleParam(matches[1], params, errors);
+  }
```
Comment (dario-piotrowicz) [inline packages/animations/browser/src/util.ts:243]: if the value is a precise `'{{ someVar }}'` then it is handled by the `interpolateSingleParam` which will convert it to a number if the respective param is, otherwise the current behaviour applies

File: packages/animations/browser/src/util.ts [hunk 7]
```diff
@@ -227,8 +231,18 @@ export function extractStyleParams(value: string|number): string[] {
 
 export function interpolateParams(
     value: string|number, params: {[name: string]: any}, errors: any[]): string|number {
+  if (typeof value === 'number') {
+    return value;
+  }
+
   const original = value.toString();
```
Comment (JoostK) [inline packages/animations/browser/src/util.ts:238]: Is this still needed now that numbers have exited through the early return? As far as the declared type (`string|number`) it should now always be a string, or is the type inaccurate in some way?

**Review by JoostK — CHANGES_REQUESTED** [review]
Thanks for taking a look at this.

I read your comment in the PR description and I'm now wondering what kind of change this is. It's more like a feature than a fix, as IIUC a number is given special treatment w.r.t. automatically adding a unit?

Regardless of whether it's a bug or feature, I do believe we do want tests for these changes. E.g. the various scenarios of single versus multiple interpolations, whitespace handling etc. It looks like your changes take care of these aspects but there is nothing to show that it behaves as desired, and perhaps more importantly it's not currently clear what benefits these changes bring.

File: packages/animations/browser/src/util.ts [hunk 8]
```diff
@@ -237,9 +251,24 @@ export function interpolateParams(
...
 
-  // we do this to assert that numeric values stay as they are
-  return str == original ? value : str;
+function interpolateSingleParam(
+    varName: string, params: {[name: string]: any}, errors: any[]): string|number {
+  let result: string|number;
+  if (!params.hasOwnProperty(varName)) {
+    // this means that the value was never overridden by the data passed in by the user
+    errors.push(`Please provide a value for the animation param ${varName}`);
+    result = '';
```
[the remark this answers is not in the document]
Comment (dario-piotrowicz) [reply packages/animations/browser/src/util.ts:262]: yes I thought the same exact thing, but didn't do it as the amount of code seems to little to grant the creation of a new function, I'll add it thanks :slightly_smiling_face:

File: packages/animations/browser/src/util.ts (hunk 7 above)
Comment (dario-piotrowicz) [reply packages/animations/browser/src/util.ts:238]: :man_facepalming:  sorry I missed that you're totally right

**Comment by dario-piotrowicz** [comment]
@JoostK thanks for reviewing the PR :slightly_smiling_face: 

I went through a bit of thinking when working on this and I have your same thoughts, when I first saw the behaviour you see in the gif in the PR's description that clearly looked like a bug to me, but as I worked on it and saw how the unit tests were handling the values (like for example `"{{ fontsize}}px"`) I realised that it is not actually a bug.

I sort of finished the changes and thought that this could bring a simplified/natural way of using the values, by using numbers directly and not being forced onto strings. But it isn't indeed a big benefit and it also changes the semantics of the strings in the animation declarations, so I am also not very sure on this either to be honest.

Since I already finished I just figured I could just keep the PR and see what you would think of it.... but I agree this may not really be a good change.... if you want I can just scrap the idea :sweat_smile:

**Comment by JoostK** [comment]
If it's unclear what the benefits are then I think closing the PR is the right thing to do at this time. Thanks again for looking into this though! Your contributions have been tremendous 💯
